drivers: serial: uart_emul: support interrupt-driven receive

cfriedt opened the pull request:
Previously, the uart_emul device did not support interrupt-driven receive, which is an unwritten requirement for hooking up a uart to the Zephyr console driver.

The console is a fairly high-traffic "device" and we should be able to perform more extensive testing on it. There were few options aside from matching twister output against a regular expression. With this change, we can use the emulation uart within the body of a ZTest.

Testing Done:
```shell
west build -p auto -b qemu_riscv64 -t run tests/drivers/console_switching
...
*** Booting Zephyr OS build zephyr-v3.4.0-4277-gae0d63471be1 ***
Running TESTSUITE console_switching
===================================================================
START - test_read
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
 PASS - test_read in 0.005 seconds
===================================================================
START - test_write
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
 PASS - test_write in 0.003 seconds
===================================================================
TESTSUITE console_switching succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [console_switching]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.008 seconds
 - PASS - [console_switching.test_read] duration = 0.005 seconds
 - PASS - [console_switching.test_write] duration = 0.003 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Fixes #63415

Commit:
drivers: serial: uart_emul: support interrupt-driven receive

Previously, the uart_emul device did not support interrupt-driven
receive, which is an unwritten requirement for hooking up a uart to
the Zephyr console driver.

The console is a fairly high-traffic subsystem, and we should be able
to perform more extensive testing on it, aside from maching twister
output against a regular expression. With this change, we can use
the emulation uart within the body of a ZTest.

Testing Done:
```shell
west build -p auto -b qemu_riscv64 -t run tests/drivers/console_switching
...
*** Booting Zephyr OS build zephyr-v3.4.0-4277-gae0d63471be1 ***
Running TESTSUITE console_switching
===================================================================
START - test_read
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
 PASS - test_read in 0.005 seconds
===================================================================
START - test_write
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
 PASS - test_write in 0.003 seconds
===================================================================
TESTSUITE console_switching succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [console_switching]: pass = 2, fail = 0, skip = 0...
 - PASS - [console_switching.test_read] duration = 0.005 seconds
 - PASS - [console_switching.test_write] duration = 0.003 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* truncated a line in the commit message  to avoid `37: UC4 Commit message body line exceeds max length (106>75): "SUITE PASS - 100.00% [console_switching]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.008 seconds"`

dcpleung commented on drivers/serial/uart_emul.c:
```diff
@@ -31,8 +38,15 @@ struct uart_emul_data {
+	void *irq_cb_user_data;
 };
 
+static void uart_emul_irq_handler(struct k_work *work);
```
I don't see it being used before the actual implementation. So remove this.

dcpleung reviewed (commented):
I think the additions can be wrapped around `CONFIG_UART_INTERRUPT_DRIVER` to minimize footprint.

Commit:
drivers: serial: uart_emul: make api implementation static

Keep uart API implementation of the uart_emul driver "private"
by adding the static keyword.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* switch from `IF_ENABLED()` to `#ifdef`
* make private API implementation `static`

cfriedt commented:
retested
```
*** Booting Zephyr OS build v3.5.0-rc1-38-g3e24c7ed6548 ***
Running TESTSUITE console_switching
===================================================================
START - test_read
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
read "Hello, uart_emul0!" from uart_emul0
read "Hello, uart_emul1!" from uart_emul1
 PASS - test_read in 0.005 seconds
===================================================================
START - test_write
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
wrote "Hello, uart_emul0!" to uart_emul0
wrote "Hello, uart_emul1!" to uart_emul1
 PASS - test_write in 0.003 seconds
===================================================================
TESTSUITE console_switching succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [console_switching]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.008 seconds
 - PASS - [console_switching.test_read] duration = 0.005 seconds
 - PASS - [console_switching.test_write] duration = 0.003 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

nick-kraus commented on drivers/serial/uart_emul.c:
```diff
@@ -99,14 +113,108 @@ int uart_emul_config_get(const struct device *dev, struct uart_config *cfg)
+{
+	struct uart_emul_data *data = dev->data;
+
+	if (!data->irq_en) {
```
This will return 1 even when there is no data in the ring buffer, which violates the uart.h API description of the function: 'Check if UART RX buffer has at least one pending character (i.e. uart_fifo_read() will succeed and return non-zero)'.

cfriedt replied:
Fixed!

cfriedt replied:
Fixed!

nick-kraus reviewed (commented):
I had just made a [similar change](https://github.com/nick-kraus/zephyr/commit/5d2a4e4378b21921649c3c7573d81bc99e60d0f5) to the UART emul driver when I noticed this pull request on the main repo. I have some code relying on the interrupt UART API and was going to use the emulation driver for testing.

I added one comment, but also wanted to mention that when building with this commit I got a Kconfig error saying UART_INTERRUPT_DRIVEN got the value 'n' because of unsatisfied dependencies. I believe that adding `select SERIAL_SUPPORT_INTERRUPT` to the Kconfig.emul file is the appropriate fix.

cfriedt commented:
> I had just made a [similar change](https://github.com/nick-kraus/zephyr/commit/5d2a4e4378b21921649c3c7573d81bc99e60d0f5) to the UART emul driver when I noticed this pull request on the main repo. I have some code relying on the interrupt UART API and was going to use the emulation driver for testing.

Cool - I see that you've added interrupt driven transmit too. That's great - would it be reasonable to add it after this?

This was originally part of #62998 and I moved it to a separate PR to narrow the scope of the other one.

> I added one comment, but also wanted to mention that when building with this commit I got a Kconfig error saying UART_INTERRUPT_DRIVEN got the value 'n' because of unsatisfied dependencies. I believe that adding `select SERIAL_SUPPORT_INTERRUPT` to the Kconfig.emul file is the appropriate fix.

Good catch, thanks.  I'll add it.

nick-kraus commented:
> Cool - I see that you've added interrupt driven transmit too. That's great - would it be reasonable to add it after this?

Yeah I think its useful to add. My version was simply synchronous calls so I can change it to use the work queue, I'd be happy to do that work once this PR lands.

I'd like to add the bit about uart error handling if you're amenable as well, helps me test some additional code paths. Once again for after this PR is merged in.

faxe1008 commented on drivers/serial/uart_emul.c:
```diff
@@ -31,9 +38,16 @@ struct uart_emul_data {
 	struct k_spinlock tx_lock;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	bool irq_en;
```
Considering that there is a change on the horizon introducing interrupt driven transmit maybe rename to `rx_irq_en`?

cfriedt replied:
Fixed!

faxe1008 commented on drivers/serial/uart_emul.c:
```diff
@@ -127,6 +235,15 @@ uint32_t uart_emul_put_rx_data(const struct device *dev, uint8_t *data, size_t s
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	if (count > 0 && drv_data->irq_en) {
+		if (!ring_buf_is_empty(drv_data->rx_rb)) {
```
Does not the check here also need to be spinlock protected?

faxe1008 commented on drivers/serial/uart_emul.c:
```diff
@@ -168,6 +285,20 @@ uint32_t uart_emul_flush_tx_data(const struct device *dev)
 
+static int uart_emul_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
```
Just for my own curiosity: The compilers that I tried with did not complain at all. Is this a common accepted pattern in zephyr to avoid a ton of macro branching?

kartben commented:
Hey @cfriedt, I think we are in feature freeze right now :)
@jhedberg @fabiobaltieri

cfriedt commented:
> Hey @cfriedt, I think we are in feature freeze right now :)
> @jhedberg @fabiobaltieri 

Sure - there are no samples being added here.

kartben commented:
Right but this definitely sounds to me like it's more than "stabilization and cosmetic changes, bug and doc fixes".
In any case, this should have probably followed the minimum review period of 2 days.

jhedberg commented:
> Sure - there are no samples being added here.

As the name suggests, feature freeze is primarily about stopping to add new features into the tree. Anyway, this PR is marked as fixing a bug, so in that respect it qualifies for merging.

cfriedt commented:
> Right but this definitely sounds to me like it's more than "stabilization and cosmetic changes, bug and doc fixes".

This was used to diagnose and fix an infinite loop in an isr (#63000) that was a problem in just about every Zephyr device in existence. #63000 was split from #62998 to make code easier to review.

#62998 itself is used to improve diagnostics and system remediation on devices that serve billions of people every day.

> In any case, this should have probably followed the minimum review period of 2 days.

Technically speaking, this code has already been already under review for 2 weeks. This was split from #62998 to make code easier to review.

I appreciate your concern, but these were far from a cosmetic changes.
